thinc/loss.py

```diff
@@ -96,26 +99,44 @@ def convert_truths(self, truths, guesses: Floats2d) -> Tuple[Floats2d, Floats2d]
                             negatives_mask[i] = 0  # type: ignore
                             negatives_mask[i][neg_index] = -1  # type: ignore
                     truths = [self._name_to_i[name] for name in truths]
+            # cast truths as 1d array
             truths = xp.asarray(truths, dtype="i")
+            # get mask for 1d array
             mask = _make_mask(guesses, missing)
         else:
+            # get mask for 2d array
```

Comment:
So `truths` is of type ``IntsOrFloatsOrStrs`, so what happens when it's `Ints1d`?

Reply:
good point, its super complicated with the handling of the different types of input for me too!

Reply:
If its `Ints1d` then it goes into `_make_mask_by_value` which handles both 1d and 2d and my comment #get mask for 2d array is lying.

Reply:
> (all these if-else branches are so confusing!)

Maybe it makes sense to break this up into multiple functions. Then `convert_truths` could become more readable. E.g.:

```
if isinstance(truths, list):
  truths = truths.truths_list_to_array(list) # Now truths is Union[Ints1d, Floats2d]

if truths.ndim == 1:
  self.truths2onehot(truths, n_classes=guesses.shape[-1]) # Not truths is Floats2d
else:
  # Apply label smoothing, etc.
```

(This is more schematic than actual code. ;) )

Reply:
Sorry, I should add that this is not necessarily for this PR. But I think we should break this up in a follow-up PR.

Reply:
> Maybe it makes sense to break this up into multiple functions. 

Yes - I was thinking the exact same thing. And yes, follow-up PR is fine :-)

Reply:
I do feel that there is this cross-entropy refactor lurking in my life. I'll add a ticket.

Reply:
Do we agree that its a next PR?

Reply:
>and my comment #get mask for 2d array is lying

This was my suspicion 😛

Reply:
I'll resolve here for now!